searx/searx
# [enh] record exception details per engine

## What does this PR do?

* Record exception details that occurs inside the engine (file, line number)
* `searx/search.py`, `send_http_request` is modified to call `response.raise_for_status()`: it makes sure the HTTP response is valid and avoids confusing error messages in the `response` function of the engine.
* ~Add an new function: `def eval_xpath_list(element, xpath_str, index=None, min_len=None, raise_on_not_found=True, default=None)`~
* Add two new functions: `eval_xpath_list(element, xpath_spec, min_len=None)` and `eval_xpath_getindex(elements, xpath_spec, index, default=NOTSET)`.
* `eval_xpath`, `eval_xpath_list` and `eval_xpath_getindex` raise `searx.exceptions.SearxXPathException` with details that are recorded
* Some engine are modified to use these functions. (see https://gist.github.com/dalf/f8dfc5878b1130be22ed8bdef3820018#file-dsl-py-L164-L183 for a more readable (?) DSL. )
* Invalid result are reported too (missing url, content or title). See `searx/results.py`, `ResultContainer.extend`
* Unexpected redirect are recorded as an error (see #2336 : most probably the engine needs to be updated).
* Errors reported to the user are simplified ("HTTP timeout", "HTTP error" and "unexpected crash").
* Add an new API `/stats/errors` (perhaps temporary, see https://github.com/searx/searx/pull/477 ).
* Exception hierarchy:
    * SearxException
      * SearxParameterException: error parsing input parameter from the user
      * SearxEngineException: error inside an engine
        * SearxXPathSyntaxException: error parsing an XPath
        * SearxEngineResponseException: error parsing the response from the website
          * SearxEngineAPIException: the website has returned an API error
          * SearxEngineCaptchaException: the website has returned a CAPTCHA
          * SearxEngineXPathException: the engine can't find a XPATH.

<details><summary>output example</summary>

```yaml
{
  "btdigg": [
    {
      "code": "response.raise_for_status()", 
      "exception_classname": "requests.exceptions.HTTPError", 
      "filename": "/home/alexandre/code/searx/searx/search.py", 
      "function": "send_http_request", 
      "line_no": 135, 
      "log_message": null, 
      "log_parameters": [
        "503", 
        "Service Unavailable", 
        "btdig.com"
      ], 
      "percentage": 100
    }
  ], 
  "google videos": [
    {
      "code": "url = eval_xpath_list(result,'.//div[@class=\"r\"]/a/@href', index=0)", 
      "exception_classname": "searx.exceptions.SearxXPathException", 
      "filename": "/home/alexandre/code/searx/searx/engines/google_videos.py", 
      "function": "response", 
      "line_no": 72, 
      "log_message": null, 
      "log_parameters": [
        ".//div[@class=\"r\"]/a/@href", 
        "Index 0 not found"
      ], 
      "percentage": 100
    }
  ], 
  "invidious": [
    {
      "code": "search_results = resp.json()", 
      "exception_classname": "json.decoder.JSONDecodeError", 
      "filename": "/home/alexandre/code/searx/searx/engines/invidious.py", 
      "function": "response", 
      "line_no": 55, 
      "log_message": null, 
      "log_parameters": [
        "Expecting value"
      ], 
      "percentage": 100
    }
  ], 
  "seedpeer": [
    {
      "code": "response = req(request_params['url'], **request_args)", 
      "exception_classname": "requests.exceptions.ConnectionError", 
      "filename": "/home/alexandre/code/searx/searx/search.py", 
      "function": "send_http_request", 
      "line_no": 134, 
      "log_message": null, 
      "log_parameters": [
        null, 
        null, 
        "seedpeer.me"
      ], 
      "percentage": 100
    }
  ], 
  "sepiasearch": [
    {
      "code": "log_error(engine_name, 'some results are invalids: ' + msg)", 
      "exception_classname": null, 
      "filename": "/home/alexandre/code/searx/searx/results.py", 
      "function": "extend", 
      "line_no": 195, 
      "log_message": "some results are invalids: invalid content", 
      "log_parameters": [], 
      "percentage": 100
    }
  ], 
  "youtube": [
    {
      "code": "results_json = loads(results_data) if results_data else {}", 
      "exception_classname": "json.decoder.JSONDecodeError", 
      "filename": "/home/alexandre/code/searx/searx/engines/youtube_noapi.py", 
      "function": "response", 
      "line_no": 54, 
      "log_message": null, 
      "log_parameters": [
        "Extra data"
      ], 
      "percentage": 100
    }
  ]
}
```

</details>


## Why is this change important?

The engines (the websites) can change at any time, it is difficult to have a feedback without extensive tests (see https://github.com/searx/searx-checker ).

This PR allows:
* an admin to check the searx instance health.
* [searx-stats2](https://github.com/dalf/searx-stats2) to collect the different errors of the public instances, and provide a global view of the engines.

## How to test this PR locally?

* make some queries that raise errors
* check `/stats/errors`
* check modified engines:
  * `!1337x`
  * `!acgsou`
  * `!apk_mirror`
  * `!archlinux`
  * `!arxiv`
  * `!bing_news`
  * `!duckduckgo_images`
  * `!elasticsearch`
  * `!google`
  * `!google_images`
  * `!google_videos` : the engine doesn't work. Check there is the same exception in the log.
  * `!ahmia` : require tor
  * `!youtube_api` : require an API key ( see https://cloud.google.com/docs/authentication/api-keys and https://console.cloud.google.com/apis/credentials )
  * XPath engines: check some XPath engines

## Author's checklist



## Related issues

Related to:
* https://github.com/searx/searx/issues/1802
* https://github.com/searx/searx/pull/477
